File: ralib/src/main/java/com/app/ralib/utils/TemporaryFileAcquirer.java
package com.app.ralib.utils;

import android.util.Log;

import com.app.ralib.Shared;

import java.io.Closeable;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Objects;

public class TemporaryFileAcquirer implements Closeable {
    private static final String TAG = "TemporaryFileAcquirer";

    private final Path preferredTempDir;
    private final ArrayList<Path> tmpFilePaths = new ArrayList<>();

    public TemporaryFileAcquirer() {
        this.preferredTempDir = Objects.requireNonNull(Shared.getContext().getExternalCacheDir())
                .toPath()
                .toAbsolutePath();
    }

    public TemporaryFileAcquirer(Path preferredTempDir) {
        this.preferredTempDir = preferredTempDir;
    }

    public Path acquireTempFilePath(String preferredSuffix) {
        Path tempFilePath = preferredTempDir.resolve(System.currentTimeMillis() + "_" + preferredSuffix);
        tmpFilePaths.add(tempFilePath);
        return tempFilePath;
    }

    public void cleanupTempFiles() {
        for (Path tmpFilePath : tmpFilePaths) {
            var isSuccessful = FileUtils.deleteDirectoryRecursively(tmpFilePath);
            if (!isSuccessful) {
                Log.w(TAG, "Failed to delete temporary file or directory: " + tmpFilePath);
            }
        }
        tmpFilePaths.clear();
    }

    @Override
    public void close() {
        cleanupTempFiles();
    }
}

File: app/src/main/java/com/app/ralaunch/core/common/util/TemporaryFileAcquirer.kt
package com.app.ralaunch.core.common.util

import android.content.Context
import com.app.ralaunch.core.logging.AppLog
import org.koin.java.KoinJavaComponent
import java.io.Closeable
import java.io.File

class TemporaryFileAcquirer : Closeable {

    private val preferredTempDir: File
    private val tmpFiles = mutableListOf<File>()

    constructor() {
        val context: Context = KoinJavaComponent.get(Context::class.java)
        preferredTempDir = requireNotNull(context.externalCacheDir)
    }

    constructor(preferredTempDir: File) {
        this.preferredTempDir = preferredTempDir
    }

    fun acquireTempFilePath(preferredSuffix: String): File {
        val tempFile = File(preferredTempDir, "${System.currentTimeMillis()}_$preferredSuffix")
        tmpFiles.add(tempFile)
        return tempFile
    }

    fun cleanupTempFiles() {
        tmpFiles.forEach { tmpFile ->
            val isSuccessful = FileUtils.deleteDirectoryRecursively(tmpFile)
            if (!isSuccessful) {
                Log.w(TAG, "Failed to delete temporary file or directory: $tmpFile")
            }
        }
        tmpFiles.clear()
    }

    override fun close() {
        cleanupTempFiles()
    }

    companion object {
        private const val TAG = "TemporaryFileAcquirer"
    }
}
